### ISSUE TITLE
Incorrect type allowed in returned object 

### ISSUE THREAD

**Issue by prostgles** [microsoft/TypeScript#39592]
#### Version 
4.0.0-beta / 3.9.5  / 3.9.2 / Nightly

##### Code
```
type V = {
    p1?: "*" | { p11: object };
    p2?: "*" | { p22: object };
}

type T = {
    p3?: "*" | { p33: object };
    p4?: "*" | { p44: object };
}

type Pt = { [key: string]: ( V | T | "*") } | "*";

type Args = { a1: any, a2: any };
type P = (a: Args) => (Pt | Promise<Pt>);

let p: P = async (a: Args) => {

    return {
        a: "*",
        b: {
            p4: 22,

            /* Only if these two are removed then p4 will fail the check */
            p1: "*",
            p2: "*"
        }
    }
};
```

#### Expected behaviour
`p4: 22` should be caught as an incorrect type

#### Actual behavior
`p4: 22` passes the type check. If the last two properties (p1 and p2) are commented out then `p4: 22` fails the type check;

#### Playground links
[3.9.2](https://www.typescriptlang.org/play/?ssl=23&ssc=76&pln=25&pc=20#code/C4TwDgpgBAalC8UDeAoK6pgIwH4BcUARAFSFQA+ymWWBA9gEYBWEAxsFAL4DcaGYAJnxFSFKoIH1mbDjxScUKUJCgAVBMj7owAZmEkylJJh06pLdl14ZMAFn2ijd2+ZlX5i5dAAKHRMYBtAGsIEAIAZ2AAJwBLADsAcwBdAgAKWDF1SgMASi4xA14lcGgAQSiE8I1jAENaKBq4kAAaBskGpvcvKG8NVJqCcsq8+AA+KFTfMW8ougBbGPCIAB5fUZyigBsIDjACXsQa8JA41gmBqCHwkfHULSgonYBXKLjNGxsLg2b7mwYCVAfIF2AgCAQ-X7AgD0xCgAHk4psQFAYgAzKDAAAWECWGIA7nQGo8HhA5nQAG4QAAmGOxbzAtigeJim02UFRNRZtOgrGxrCCUGIUMhQOwBG+Io+gnFpElChsCh4QA)

**Comment by RyanCavanaugh** [comment]
This is a duplicate of [#36945](https://github.com/microsoft/TypeScript/issues/36945). The inner object is assignable to the open `V` type through `p1: "*"`; it does not need to satisfy `T`. When a fresh object literal is checked against `V | T`, excess-property checking considers property names from all union members. Since `p4` exists in `T`, it is not an excess property merely because `22` is not valid for `T["p4"]`.
